libirods_rule_engine_plugin-audit_amqp.cpp

```diff
@@ -1,443 +1,542 @@
-                key.c_str(),
-                json_string(elem.second.c_str()));
-
-            ++ctr; 
-            ctr_str.clear();
-
-        } // for elem
-    } // for itr
-
-    char* tmp_buf = json_dumps( obj, JSON_INDENT( 0 ) );
-    std::string msg_str = std::string("__BEGIN_JSON__") + std::string(tmp_buf) + std::string("__END_JSON__");
-
-    pn_message_t * message;
-    pn_data_t * body;
-
-    message = pn_message();
-
-    std::string address = audit_amqp_location + "/" + audit_amqp_topic;
-
-    pn_message_set_address(message, address.c_str());
-    body = pn_message_body(message);
-    pn_data_put_string(body, pn_bytes(msg_str.length(), msg_str.c_str()));
-    pn_messenger_put(messenger, message);
-    pn_messenger_send(messenger, -1);
-    
-    pn_message_free(message);
-
-    free(tmp_buf);
-    json_decref(obj);
-
-    if (test_mode) {
-        log_file_ofstream << msg_str << std::endl;
-    }
-
-    return err;
-}
+#include <boost/asio/ip/host_name.hpp>
+#include <boost/config.hpp>
+
+// proton-cpp includes
+#include <proton/connection_options.hpp>
+#include <proton/container.hpp>
+#include <proton/message.hpp>
+#include <proton/messaging_handler.hpp>
+#include <proton/tracker.hpp>
+#include <proton/sender.hpp>
+
+namespace
+{
+	const auto pep_regex_flavor = std::regex::ECMAScript;
+
+	// NOLINTBEGIN(cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables)
+	const std::string_view default_pep_regex_to_match{"audit_.*"};
+	const std::string_view default_amqp_topic{"irods_audit_messages"};
+	const std::string_view default_amqp_location{"localhost:5672"};
+	const std::string_view default_amqp_user;
+	const std::string_view default_amqp_password;
+	const std::string_view default_amqp_options;
+
+	std::string audit_pep_regex_to_match{default_pep_regex_to_match};
+	std::string audit_amqp_topic{default_amqp_topic};
+	std::string audit_amqp_location{default_amqp_location};
+	std::string audit_amqp_user{default_amqp_user};
+	std::string audit_amqp_password{default_amqp_password};
+	std::string audit_amqp_options{default_amqp_options};
+
+	std::regex audit_pep_regex{audit_pep_regex_to_match, pep_regex_flavor};
+
+	std::mutex audit_plugin_mutex;
+	// NOLINTEND(cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables)
+
+	const char* const rule_engine_name = "audit_amqp";
+
+	using log_re = irods::experimental::log::rule_engine;
+
+#if __cpp_lib_chrono >= 201907
+	// use utc_clock if it's implemented (for leap seconds)
+	using ts_clock = std::chrono::utc_clock;
+#else
+	// fallback to system_clock
+	using ts_clock = std::chrono::system_clock;
+#endif
+
+	// See qpid-cpp docs
+	// https://qpid.apache.org/releases/qpid-proton-0.36.0/proton/cpp/api/simple_send_8cpp-example.html
+	class send_handler : public proton::messaging_handler
+	{
+	  private:
+		const std::string& amqp_location;
+		const std::string& amqp_topic;
+		const std::string& user;
+		const std::string& password;
+		const std::string& message;
```

Comment:
Is it safe for these to be references?

Reply:
Yes.
Aside from `message`, all of these come from the global configuration variables. The only thing that changes these is `get_re_configs()`, which is only called from `start()`.
At present, `send_handler` objects only live long enough to connect, send a single message, then disconnect. The only functions to construct `send_handler` objects are `start()`, `stop()`, and `exec_rule()`, all of which immediately lock `audit_plugin_mutex`.
Therefore, there is no possibility of the configuration variables being changed while a `send_handler` object is alive.

Reply:
I'm more so concerned with the following snippet leading to dangling references and potentially crashing the server.
```c++
send_handler handler(
￼  msg_str,
￼  audit_amqp_location,
￼  audit_amqp_topic,
￼  audit_amqp_user,
￼  audit_amqp_password);
proton::container(handler).run();
```
Is the message guaranteed to be sent by the time `.run()` returns?

Reply:
To my understanding, yes.

Reply:
Good.

Reply:
Just double-checked.
From the [API documentation for `proton::container::run()'](https://qpid.apache.org/releases/qpid-proton-0.36.0/proton/cpp/api/classproton_1_1container.html#a13a43e6d814de94978c515cb084873b1):
> Run the container in the current thread. 
> The call returns when the container stops.

Reply:
Awesome